### DISCUSSION TITLE
Hook for serve redirect from S3

### DISCUSSION THREAD

**Discussion by thangngoc89** [pocketbase/pocketbase#4871] [Show and tell]
Given that my application is mainly dealing with serving large files, proxying the files from S3 -> pocketbase -> client adds huge latency and consume a lot of pocketbase's VPS bandwidth. 

I've read several discussions where @ganigeorgiev said about not supporting serving directly from S3 due to inconsistency between S3-compatible services and that's very reasonable. 

So here is the hook that you could use for redirecting from Pocketbase's `/api/files/...` to `s3` url. This wouldn't work if the client doesn't follow redirect status code. 

Any suggestions for improvement are welcomed since I've only picked up Go for 2 weeks in order to work with Pocketbase

See updated code in

https://github.com/pocketbase/pocketbase/discussions/4871#discussioncomment-9404120

Below are kept for reference.

---

```golang
package main

import (
	"time"
        "github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"

       "github.com/pocketbase/pocketbase"
       "github.com/pocketbase/pocketbase/core"
       "github.com/pocketbase/pocketbase/hooks"
)

// Modified from Pocketbase's codebase
func NewS3(
	bucketName string,
	region string,
	endpoint string,
	accessKey string,
	secretKey string,
	s3ForcePathStyle bool,
) (*s3.Client, error) {
	ctx := context.Background() // default context

	cred := credentials.NewStaticCredentialsProvider(accessKey, secretKey, "")

	cfg, err := config.LoadDefaultConfig(ctx,
		config.WithCredentialsProvider(cred),
		config.WithRegion(region),
		config.WithEndpointResolverWithOptions(aws.EndpointResolverWithOptionsFunc(func(service, region string, options ...interface{}) (aws.Endpoint, error) {
			// ensure that the endpoint has url scheme for
			// backward compatibility with v1 of the aws sdk
			prefixedEndpoint := endpoint
			if !strings.Contains(endpoint, "://") {
				prefixedEndpoint = "https://" + endpoint
			}

			return aws.Endpoint{URL: prefixedEndpoint, SigningRegion: region}, nil
		})),
	)
	if err != nil {
		return nil, err

	}

	client := s3.NewFromConfig(cfg, func(o *s3.Options) {
		o.UsePathStyle = s3ForcePathStyle
	})

	return client, nil
}
func main() {
    app := pocketbase.New()

    app.OnFileDownloadRequest().Add(func(e *core.FileDownloadEvent) error {
		s3settings := app.Settings().S3
		if s3settings.Enabled {
			s3client, err := NewS3(s3settings.Bucket, s3settings.Region, s3settings.Endpoint, s3settings.AccessKey, s3settings.Secret, s3settings.ForcePathStyle)
			if err != nil {
				app.Logger().Error("create s3 client", err)
				return nil
			}
			presignClient := s3.NewPresignClient(s3client)
			presignedUrl, err := presignClient.PresignGetObject(context.Background(),
				&s3.GetObjectInput{
					Bucket: aws.String(s3settings.Bucket),
					Key:    aws.String(e.ServedPath),
				},
				s3.WithPresignExpires(time.Hour))
			if err != nil {
				app.Logger().Error("create s3 presigned url", err)
				return nil
			}
			e.HttpContext.Redirect(http.StatusTemporaryRedirect, presignedUrl.URL)
			return hook.StopPropagation
		}

		return nil
	})

    if err := app.Start(); err != nil {
        log.Fatal(err)
    }
}
```

**Comment by pnmcosta** [comment]
Noice. One slight optimization I notice, I don't think you need to instance the `s3client` for each request (but I could be wrong, I don't actively use S3).

You could do so only once for the app lifecycle on the https://pocketbase.io/docs/go-event-hooks/#onafterbootstrap event (which would have loaded the settings) and then just check if the client exists on the OnFileDownloadRequest handler, instead of recreating it for all incoming requests.

**Reply by thangngoc89** [reply]
You're absolutely right. I've made some improvements in my project but forgot to update it here. This is the `OnAfterBootstrap` hook

```go
func main() {
	var appS3Client *s3.Client
	var appS3Bucket *string
        
        app := pocketbase.New()
	app.OnAfterBootstrap().Add(func(e *core.BootstrapEvent) error {
		s3client, bucketName, err := NewS3FromSettings(e.App.Settings().S3)
		if err != nil {
			app.Logger().Error("failed to initialize s3 client serving directly from /api/files", err)
			return nil
		}
		appS3Client = s3client
		appS3Bucket = bucketName
		return nil
	})
       
        app.OnFileDownloadRequest().Add(func(e *core.FileDownloadEvent) error {
                // IMPORTANT: adjust your conditions here. 
                // I wanted to keep the lazy thumbnail generation of Pocketbase for images while redirect for large files
		if appS3Client != nil && (strings.HasSuffix(e.ServedPath, ".7z") || strings.HasSuffix(e.ServedPath, ".zip")) {
			presignClient := s3.NewPresignClient(appS3Client)
			presignedUrl, err := presignClient.PresignGetObject(context.Background(),
				&s3.GetObjectInput{
					Bucket: appS3Bucket,
					Key:    aws.String(e.ServedPath),
				},
				s3.WithPresignExpires(time.Hour))
			if err != nil {
				app.Logger().Error("create s3 presigned url", err)
				return nil
			}
			e.HttpContext.Redirect(http.StatusMovedPermanently, presignedUrl.URL)
			return hook.StopPropagation
		}

		return nil
	})
}

func newS3(
	// bucketName string,
	region string,
	endpoint string,
	accessKey string,
	secretKey string,
	s3ForcePathStyle bool,
) (*s3.Client, error) {
	ctx := context.Background() // default context

	cred := credentials.NewStaticCredentialsProvider(accessKey, secretKey, "")

	cfg, err := config.LoadDefaultConfig(ctx,
		config.WithCredentialsProvider(cred),
		config.WithRegion(region),
		config.WithEndpointResolverWithOptions(aws.EndpointResolverWithOptionsFunc(func(service, region string, options ...interface{}) (aws.Endpoint, error) {
			// ensure that the endpoint has url scheme for
			// backward compatibility with v1 of the aws sdk
			prefixedEndpoint := endpoint
			if !strings.Contains(endpoint, "://") {
				prefixedEndpoint = "https://" + endpoint
			}

			return aws.Endpoint{URL: prefixedEndpoint, SigningRegion: region}, nil
		})),
	)
	if err != nil {
		return nil, err

	}

	client := s3.NewFromConfig(cfg, func(o *s3.Options) {
		o.UsePathStyle = s3ForcePathStyle
	})

	return client, nil
}

func NewS3FromSettings(s3settings settings.S3Config) (*s3.Client, *string, error) {
	if s3settings.Enabled {
		s3client, err := newS3(s3settings.Region, s3settings.Endpoint, s3settings.AccessKey, s3settings.Secret, s3settings.ForcePathStyle)
		if err != nil {
			return nil, nil, err
		}
		bucket := aws.String(s3settings.Bucket)
		return s3client, bucket, nil
	}
	return nil, nil, nil
}
```

**Reply by pnmcosta** [reply]
Awesome. One slight nitpick, in GO it is preferable to use zero values on integral types rather than pointers, it is easier and safer to check down path.

The `var appS3Bucket *string` can simply be `var appS3Bucket string` and you can return `""` from the `NewS3FromSettings`, this means you can just check for:

```
if appS3Bucket != ""{
}
```

There are exceptions of course (but I don't think it's the case here), have a look at https://dhdersch.github.io/golang/2016/01/23/golang-when-to-use-string-pointers.html which explains it in much more depth than I have here.

Happy coding and have a great Sunday :)

**Reply by thangngoc89** [reply]
Thanks a lot for nitpicking. Like I said 3 weeks experience with go. 
Happy coding to you too.

**Reply by thangngoc89** [reply]
@pnmcosta if possible, please upvote this issue for more attentions. 
https://github.com/pocketbase/pocketbase/issues/4878

It pairs nicely with this hook and I'm running a fork of pocketbase with the modification

**Reply by pnmcosta** [reply]
Umm, I've had a look at it - but I have very little S3 experience - but abstractly thinking about it, I don't think that would ever be done, simply because it's a specific need for your use case that would probably not be beneficial to all PB users - but I could be totally wrong about this 😅

Just a thought, since PB is very extensible, have you looked at whether you could use any events to override the default upload behaviour? Even if you have to instance your own client/blob/whatnot instead of using PB's default to upload?

**Reply by thangngoc89** [reply]
> Just a thought, since PB is very extensible, have you looked at whether you could use any events to override the default upload behaviour? Even if you have to instance your own client/blob/whatnot instead of using PB's default to upload?

Of course. I investigated about it a little it and it's totally possible. But it's way down in the never ending todo list. See https://github.com/pocketbase/pocketbase/discussions/4858

> I don't think that would ever be done

Yup. I thought so too but doing this doesn't actually change any functionality of Pocketbase. It wouldn't affect files serving through `/api/files`. And as stated in the proposal, Pocketbase set the correct `Cache-Control` header at serving time so why not set it at writing time too.

**Reply by pnmcosta** [reply]
I think you might be able to, check this out:

> Once added, you could create/update a Record and upload "documents" files by sending a multipart/form-data request using the Records create/update APIs.

from https://pocketbase.io/docs/files-handling/

Which means you could handle https://pocketbase.io/docs/go-event-hooks#onrecordaftercreaterequest (or update) and custom upload using your own adapter/writer, and reset `e.UploadedFiles` to nil/empty so the default upload does not occur.

But there could be things I'm not considering (are aware of) like the reference to the files on the record, etc.

**Reply by thangngoc89** [reply]
You're right. I could avoid forking Pocketbase (even though it's literally 4 lines of code) by listening to the record create/update hook and set the cache-control headers for new files myself. Longer than 4 LOC but easier in the long run. I'll post the code once I've done it in the issue

**Reply by pnmcosta** [reply]
> See https://github.com/pocketbase/pocketbase/discussions/4858

Oh, there you go, you can effectively have your own tunned up upload process

> Pocketbase set the correct Cache-Control header at serving time so why not set it at writing time too.

Although the name is similar, I guess a response Cache-Control header is different from the Cache-Control writing to the bucket, also consider that not all users might be using the same S3/bucket provider (I believe this to be supported, but not 100% sure) and also you could have writes for different types of files and cache requirements. Again, I could be totally wrong on my assumptions so take them with a pinch of salt :D

**Reply by thangngoc89** [reply]
> a response Cache-Control header is different from the Cache-Control writing to the bucket

In term of S3, you write `Cache-Control` header or any valid HTTP headers at file creation time or update it at some point, and the S3 will serve those exact same headers back when you request it. Maybe with the exception of CORS header. (not so sure about the CORS header part).

>  not all users might be using the same S3/bucket provider 

This is true. I've only used a few S3 providers. But considering that this is in the `s3blob` abstraction library. When you upload to local file system, it has a `filename.attr` for storing uploaded header. 

> also you could have writes for different types of files and cache requirement

Not true for pocketbase since every uploaded file has a unique name if you use the official API. This is why `/api/files` serves files with forever cache-control header

**Reply by pnmcosta** [reply]
> In term of S3, you write Cache-Control header or any valid HTTP headers at file creation time or update it at some point, and the S3 will serve that exact same headers back when you request it. Maybe with the exception of CORS header. (not so sure about the CORS header part).

Oh I see, that makes sense, so essentially at write time you're telling S3 what the response headers for the blob will be. Got it, great info to know.

> This is why /api/files serves files with forever cache-control header

Yes, but not only does it serve S3 blobs, but also local filesystem files, I'm thinking that's to prevent server disk IO if the client has cached it. 

Perhaps, serving a remote blob should observe whatever has been set for its response headers instead, and then I think extending the upload to customize this could make sense (but again I may not be considering all use cases).

I'm still not convinced that all files should be uploaded with the same Cache-Control header value, you could have collections with ephemeral files and others that are not. But it could be part of the collection config to specify as such 🤔

**Reply by thangngoc89** [reply]
> Yes, but not only does it serve S3 blobs, but also local filesystem files

Yes. And in both cases, it would overwrite the Cache Control header with.

https://github.com/pocketbase/pocketbase/blob/5acaa0a55c9f2d5070101857543c746cac823c80/tools/filesystem/filesystem.go#L395

> I'm thinking that's to prevent server disk IO if the client has cached it.

Yes. And it could also prevent transmission completely if the browser already has this in cache. so save bandwidth.

> I'm still not convinced that all files should be uploaded with the same Cache-Control header value, you could have collections with ephemeral files and others that are not. But it could be part of the collection config to specify as such 

Pocketbase is already doing this for all files by setting Cache-Control header value when serving `/api/files`. 

My issue is about setting the same header when saving so that when I redirect to S3, it would have the same header and I get the same caching behavior when serving from `/api/files` without eating the VPS's bandwidth. 

So my proposal doesn't actually change any of Pocketbase's public API/behaviors.
